# НИС "Анализ данных в Python"

# Задание

Необходимо проанализировать данные о сотрудниках компании (файл employees.xlsx). 

В файле содержаться следующие переменные:

   - id - идентификатор сотрудника
   
   
   - gender - пол: 'm' (мужской), 'f' (женский)
   
      
   - marital - семейное положение: 'divorced', 'married', 'single', 'unknown'
    
   
   - education - число лет обучения в годах
   
   
   - jobcat - категория работника: 'Programmer', 'Manager', 'Technician'
   
   
   - salary - текущая заработная плата (в условных единицах)
   
   
   - salbegin - зарплата при приёме на работу (в условных единицах)
   
   
   - jobtime - число месяцев, прошедших с даты найма
   
   
   - prevexp - общий опыт работы в месяцах
   
  
   - minority - принадлежность к национальным меньшинствам: 'Yes', 'No'
 



### ! Важное уточнение: для того, чтобы пункт задания был засчитан, необходимо его выполнить полностью, не только провести расчёты, но и сформулировать выводы по полученным результатам (где требуется). Все пункты задания независимы друг от друга. Каждый пункт нужно выполнять с исходными данными.

1. Охарактеризуйте размер датасета и полноту данных в нём. Какие типы колонок в датафрейме?

In [49]:
#здесь написать ответ
import pandas as pd
df = pd.read_excel('employees.xlsx')
df

id gender   marital  education      jobcat   salary  salbegin  jobtime  \
0      1      m   married         15     Manager  57000.0   27000.0       98   
1      2      m   married         16  Programmer  40200.0   18750.0       98   
2      3      f   married         12  Programmer  21450.0   12000.0       98   
3      4      f   married          8  Programmer  21900.0   13200.0       98   
4      5      m   married         15  Programmer  45000.0   21000.0       98   
..   ...    ...       ...        ...         ...      ...       ...      ...   
469  470      m   married         12  Programmer  26250.0   15750.0       64   
470  471      m    single         15  Programmer  26400.0   15750.0       64   
471  472      m   married         15  Programmer  39150.0   15750.0       63   
472  473      f  divorced         12  Programmer  21450.0   12750.0       63   
473  474      f   married         12  Programmer  29400.0   14250.0       63   

     prevexp minority  
0      144.0       No  
1       36.0       No  
2      381.0       No  
3      190.0       No  
4      138.0       No  
..       ...      ...  
469     69.0      Yes  
470     32.0      Yes  
471     46.0       No  
472    139.0       No  
473      9.0       No  

[474 rows x 10 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 474 entries, 1 to 474
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   gender     474 non-null    object 
 1   marital    474 non-null    object 
 2   education  474 non-null    int64  
 3   jobcat     471 non-null    object 
 4   salary     467 non-null    float64
 5   salbegin   470 non-null    float64
 6   jobtime    474 non-null    int64  
 7   prevexp    450 non-null    float64
 8   minority   474 non-null    object 
dtypes: float64(3), int64(2), object(4)
memory usage: 37.0+ KB


In [5]:
df.jobcat.value_counts()

jobcat
Programmer    361
Manager        83
Technician     27
Name: count, dtype: int64

474 строки, 10 столбцов, несколько столбцов содержат пропуски (указать процент).
Тип шкалы измерения значений:
- номинальная (gender, minority, marital, jobcat)
- порядковая (нет)
- метрическая (education и все остальные)

2. Рассчитайте статистические показатели для переменных, имеющих метрическую шкалу измерения. Прокомментируйте на основе полученных расчётов наблюдаемые тенденции в распределении значений рассмотренных переменных. 

In [11]:
#здесь написать ответ
df.describe()

education         salary      salbegin     jobtime     prevexp
count  474.000000     467.000000    470.000000  474.000000  450.000000
mean    13.491561   34190.845824  17054.627660   81.109705  100.973333
std      2.884846   16789.484900   7890.359771   10.060945  104.907443
min      8.000000   15750.000000   9000.000000   63.000000    2.000000
25%     12.000000   24000.000000  12637.500000   72.000000   24.000000
50%     12.000000   28800.000000  15000.000000   81.000000   59.000000
75%     15.000000   36375.000000  17872.500000   90.000000  144.000000
max     21.000000  135000.000000  79980.000000   98.000000  476.000000

3. Проведите частотный анализ категориальных переменных. Выведите число уникальных значений и самое часто встречаемое значение. Охарактеризуйте распределение значений, как в абсолютном, так и в процентном отношении. 

In [8]:
#здесь написать ответ
df.marital.value_counts()

marital
married     334
single       85
divorced     52
unknown       3
Name: count, dtype: int64

In [12]:
df.marital.unique()

array(['married', 'single', 'divorced', 'unknown'], dtype=object)

In [13]:
df.marital.nunique()

4

In [14]:
df.marital.value_counts(normalize=True)

marital
married     0.704641
single      0.179325
divorced    0.109705
unknown     0.006329
Name: proportion, dtype: float64

4. Примените многоуровневую сортировку.

In [18]:
#здесь написать ответ
#по education по возрастанию, по salary по убыванию
df.sort_values(by=['education', 'salary'], ascending = [True, False]).tail(25)

gender   marital  education      jobcat   salary  salbegin  jobtime  \
id                                                                        
66       m   married         19         NaN  78125.0   30000.0       93   
235      m   married         19     Manager  75000.0   31500.0       81   
456      m   married         19     Manager  75000.0   42510.0       65   
284      m   married         19     Manager  73500.0   33000.0       78   
198      m    single         19     Manager  70875.0   43500.0       83   
420      m    single         19     Manager  70000.0   35040.0       67   
89       m   married         19     Manager  68750.0   27480.0       92   
283      m   married         19     Manager  68125.0   32010.0       78   
389      m    single         19     Manager  66875.0   32490.0       69   
430      m   married         19     Manager  66250.0   34980.0       67   
387      m   married         19     Manager  65000.0   31980.0       69   
257      m   married         19     Manager  61875.0   36750.0       80   
458      m    single         19     Manager  61875.0   28740.0       65   
408      m   married         19     Manager  61250.0   33000.0       68   
27       m   married         19     Manager  60375.0   27480.0       96   
232      m  divorced         19     Manager  55500.0   33750.0       81   
450      m   married         19     Manager  55000.0   34980.0       65   
256      m   married         19     Manager  52125.0   27480.0       80   
464      m    single         19     Manager  47550.0   33000.0       64   
409      m   married         19  Programmer  36000.0   19500.0       68   
34       m   married         19     Manager      NaN   39990.0       96   
88       m   married         19     Manager      NaN   28740.0       92   
173      m   married         20     Manager  69250.0   42480.0       85   
130      m   married         20     Manager  59375.0   30000.0       89   
137      m   married         21     Manager  65000.0   37500.0       88   

     prevexp minority  
id                     
66       7.0       No  
235     13.0       No  
456     54.0       No  
284     45.0       No  
198    156.0       No  
420     75.0       No  
89       8.0       No  
283     35.0       No  
389     81.0       No  
430     99.0      Yes  
387     74.0       No  
257    199.0       No  
458     26.0       No  
408      9.0       No  
27      96.0       No  
232     62.0       No  
450    129.0       No  
256    221.0       No  
464     27.0       No  
409     21.0       No  
34     175.0       No  
88      10.0       No  
173    134.0       No  
130      6.0       No  
137    264.0       No

5. Сколько в выборке программистов? А женщин? Какой процент в выборке женщин-программистов?

In [22]:
#здесь написать ответ
pd.pivot_table(df, index='jobcat', columns='gender', aggfunc='size', fill_value=0)

gender        f    m
jobcat              
Manager      10   73
Programmer  205  156
Technician    0   27

6. Вставьте вместо пропусков в колонку jobcat самое частое значение в этой колонке.

In [ ]:
#здесь написать ответ

7. Сравните средние зарплаты мужчин и женщин на разных должностях. Прокомментируйте полученные результаты.

In [30]:
#здесь написать ответ
df.groupby(['jobcat', 'gender'])['salary'].mean()

jobcat      gender
Manager     f         47213.500000
            m         65508.928571
Programmer  f         25045.882353
            m         31590.454545
Technician  m         30946.153846
Name: salary, dtype: float64

8. Создайте и выведите новый датафрейм, в который войдут 10 мужчин-сотрудников с самой высокой заработной платой при приёме на работу. В нем строки должны быть отсортированы по заработной плате в порядке убывания, а индексом должны являться числа от 0 до 9 включительно. Оставьте в новом датафрейме (помимо индекса) только колонки id, gender, education, salary, salbegin

In [51]:
sorted_df = df[df["gender"] == "m"].sort_values(by='salary', ascending=False)
sorted_df = sorted_df[:10]
sorted_df['ind'] = range(10)
sorted_df = sorted_df[['id', 'gender', 'education', 'salary', 'salbegin']]
sorted_df

id gender  education    salary  salbegin
28    29      m         19  135000.0   79980.0
31    32      m         19  110625.0   45000.0
17    18      m         16  103750.0   27510.0
342  343      m         16  103500.0   60000.0
445  446      m         16  100000.0   44100.0
102  103      m         19   97000.0   35010.0
105  106      m         19   91250.0   29490.0
453  454      m         19   90625.0   31250.0
430  431      m         18   86250.0   45000.0
273  274      m         16   83750.0   21750.0

In [45]:
result_df

Empty DataFrame
Columns: [id, gender, education, salary, salbegin]
Index: []

9. Между какими двумя столбцами исходного датафрейма наблюдается самая сильная корреляция?

In [33]:
#здесь написать ответ
df.corr(numeric_only=True)

education    salary  salbegin   jobtime   prevexp
education   1.000000  0.654532  0.633266  0.047379 -0.286943
salary      0.654532  1.000000  0.877525  0.067227 -0.140432
salbegin    0.633266  0.877525  1.000000 -0.015768  0.010077
jobtime     0.047379  0.067227 -0.015768  1.000000  0.001925
prevexp    -0.286943 -0.140432  0.010077  0.001925  1.000000

10. Создайте переменную, которая бы показывала, увеличилась ли у сотрудника зарплата (1) или нет (0). 

In [ ]:
#здесь написать ответ

11. Перекодируйте переменную salary в переменную с порядковой шкалой, отражающую зарплатную группу, к которой относится сотрудник (должно быть не менее 5 групп). Постройте частотную таблицу по новой переменной. Прокомментируйте полученные результаты.

In [ ]:
#здесь написать ответ

12. Постройте сводные таблицы (не менее двух таблиц). Прокомментируйте полученные результаты, какие тенденции в распределении значений переменных они отражают?

In [ ]:
#здесь написать ответ

13. Сравните внешний вид распределения зарплат сотрудников на разных должностях. Опишите как эти распределения отличаются от нормального распределения.  Сделайте выводы.

In [ ]:
#здесь написать ответ


14. Постройте столбиковую и круговую диаграммы по категориальным переменным. Постройте столбиковую диаграмму с накоплением (stacked). Прокомментируйте наблюдаемые на графиках тенденции.

In [ ]:
#здесь написать ответ

15. Сформулируйте не менее 10 исследовательских гипотез. Подберите подходящие методы анализа данных для их проверки. Какие из них нашли подтверждение, а какие были опровергнуты по результатам анализа данных и почему?

#здесь написать ответ
H1: зарплаты мужчини и женщин на разных должностях не отличаются (отклоняем)
H2: есть прямая полодительная зависимость между опытом работы и зарплатой (принимаем)
    